### PR TITLE
Add com.microsoft.Teams

### PR DESCRIPTION
**Pull request by Lctrs**

### CODE REVIEW

**Comment by Eonfge** [comment]
bot, build com.microsoft.teams.Teams

Related, it would be good to migrate all existing from this app:
https://github.com/flathub/com.github.IsmaelMartinez.teams_for_linux

**Comment by Lctrs** [comment]
> Related, it would be good to migrate all existing from this app:
> https://github.com/flathub/com.github.IsmaelMartinez.teams_for_linux

@Eonfge don't know what you mean by this ?

**Comment by Lctrs** [comment]
If you mean migrating account information, settings, etc., I'm not sure that's feasible.

`com.microsoft.teams.Teams` is a brand new app, it's not a new rebranded version of `com.github.IsmaelMartinez.teams_for_linux`.

**Comment by Eonfge** [comment]
@Lctrs No, this was a comment to the maintainers of Flathub. It's possible for Flathub maintainers to declare an application End of Life, after which they can instruct Flatpak to automatically download a new version. In this case, I recommend them to declare the old wrapped EoL, and to automatically move all those users to this new version.

**Comment by Lctrs** [comment]
Oh okay. Sorry for the confusion.

In this case, I'm not sure we should do it right now. The new app by Microsoft is still in an alpha/beta stage.

File: flathub.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "publish-delay-hours": 0,
+  "only-arches": ["x86_64"],
+  "skip-appstream-check": true
```
Comment (nedrichards) [inline flathub.json]: Why is this here? It should not be.

File: com.microsoft.teams.Teams.json [hunk 2]
```diff
@@ -0,0 +1,106 @@
+{
+  "id": "com.microsoft.teams.Teams",
```
Comment (nedrichards) [inline com.microsoft.teams.Teams.json]: Is this the app-id of the app itself or did you suggest this?

**Comment by nedrichards** [comment]
@Lctrs have you contacted upstream and asked if they wish to maintain the app on flathub?

File: com.microsoft.teams.Teams.json (hunk 2 above)
Comment (Lctrs) [reply com.microsoft.teams.Teams.json]: I suggest it. Couldn't find any appdata in the base `.deb`.

**Comment by Lctrs** [comment]
@nedrichards No and I doubt Microsoft will ever answer honestly.

**Comment by LeandroStanger** [comment]
bot, build com.microsoft.teams.Teams

**Comment by LeandroStanger** [comment]
bot, build com.microsoft.teams.Teams

File: com.microsoft.teams.Teams.json [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  "finish-args": [
+    "--share=ipc",
+    "--share=network",
+    "--socket=x11",
+    "--socket=pulseaudio",
+    "--device=all",
+    "--filesystem=home",
+    "--talk-name=org.freedesktop.Notifications",
+    "--talk-name=org.kde.StatusNotifierWatcher",
+    "--talk-name=org.gtk.Notifications"
```
Comment (TingPing) [inline com.microsoft.teams.Teams.json]: I doubt this is used.

File: teams.sh [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env sh
+
+set -e
+
+exec env TMPDIR=$XDG_CACHE_HOME \
```
Comment (TingPing) [inline teams.sh]: You can do `--env=TMDPR=/var/tmp` in finish args to avoid this wrapper script.

**Comment by TingPing** [comment]
> No and I doubt Microsoft will ever answer honestly.

That is fine but we at least try.

File: teams.sh (hunk 4 above)
Comment (Lctrs) [reply teams.sh]: I prefer to rely on the wrapper script :)

**Comment by barthalion** [comment]
I have contacted Christopher Scott before this PR was opened and he promised to forward my question to product owners.

File: com.microsoft.teams.Teams.json (hunk 2 above)
Comment (barthalion) [reply com.microsoft.teams.Teams.json]: com.microsoft.Teams would be a better choice.

File: com.microsoft.teams.Teams.json [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    "proprietary"
+  ],
+  "separate-locales": false,
+  "finish-args": [
+    "--share=ipc",
+    "--share=network",
+    "--socket=x11",
+    "--socket=pulseaudio",
+    "--device=all",
+    "--filesystem=home",
```
Comment (barthalion) [inline com.microsoft.teams.Teams.json]: Limit it to downloads please.

File: com.microsoft.teams.Teams.json [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+      "build-commands": [
+        "install -Dm0755 /usr/lib/$(gcc -print-multiarch)/libbfd-*.so -t ${FLATPAK_DEST}/lib/",
+        "install -Dm0755 $(which ar) ${FLATPAK_DEST}/bin/"
+      ]
+    },
+    {
+      "name": "desktop-file-edit",
+      "buildsystem": "simple",
+      "build-commands": [
+        "install -Dm0755 $(which desktop-file-edit) ${FLATPAK_DEST}/bin/"
```
Comment (barthalion) [inline com.microsoft.teams.Teams.json]: Can you just put it into build commands of Teams?

File: com.microsoft.teams.Teams.json [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    "--filesystem=home",
+    "--talk-name=org.freedesktop.Notifications",
+    "--talk-name=org.kde.StatusNotifierWatcher",
+    "--talk-name=org.gtk.Notifications"
+  ],
+  "modules": [
+    {
+      "name": "ar",
+      "buildsystem": "simple",
+      "build-commands": [
```
Comment (barthalion) [inline com.microsoft.teams.Teams.json]: Can you just put it into build commands of Teams?

File: com.microsoft.teams.Teams.json [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+          "path": "teams.sh"
+        },
+        {
+          "type": "file",
+          "path": "com.microsoft.teams.Teams.metainfo.xml"
+        },
+        {
+          "type": "extra-data",
+          "filename": "teams.deb",
+          "only-arches": [
```
Comment (barthalion) [inline com.microsoft.teams.Teams.json]: No need to define it, flathub.json already limits target architectures.

File: apply_extra.sh [hunk 9]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env sh
```
Comment (barthalion) [inline apply_extra.sh]: Most if not all extra-data applications inline it into manifest.

File: teams.sh (hunk 4 above)
Comment (barthalion) [reply teams.sh]: Most if not all extra-data applications inline it into manifest.

File: com.microsoft.teams.Teams.json (hunk 8 above)
Comment (Lctrs) [reply com.microsoft.teams.Teams.json]: It's required by the external data checker.

Comment (barthalion) [reply com.microsoft.teams.Teams.json]: Only for Flash checker.

Comment (Lctrs) [reply com.microsoft.teams.Teams.json]: [Nope](https://github.com/flathub/flatpak-external-data-checker/blob/master/src/checkers/debianrepochecker.py#L109).

File: com.microsoft.teams.Teams.json (hunk 2 above)
Comment (Lctrs) [reply com.microsoft.teams.Teams.json]: Okay will rename.

File: com.microsoft.teams.Teams.json (hunk 8 above)
Comment (barthalion) [reply com.microsoft.teams.Teams.json]: It should be changed to default to x86_64. I may take a look at it today.

File: apply_extra.sh (hunk 9 above)
Comment (Lctrs) [reply apply_extra.sh]: I prefer to keep it separated. More maintainable for me than having it hidden as a bunch of strings inside a json file.

File: teams.sh (hunk 4 above)
Comment (Lctrs) [reply teams.sh]: I prefer to keep it separated. More maintainable for me than having it hidden as a bunch of strings inside a json file.

File: com.microsoft.teams.Teams.json (hunk 6 above)
Comment (Lctrs) [reply com.microsoft.teams.Teams.json]: I prefer to keep it separated. For me compiling or installing a dependency from the SDK is the same and should be kept in its own module.

File: com.microsoft.teams.Teams.json (hunk 7 above)
Comment (Lctrs) [reply com.microsoft.teams.Teams.json]: I prefer to keep it separated. For me compiling or installing a dependency from the SDK is the same and should be kept in its own module.

File: com.microsoft.teams.Teams.json (hunk 5 above)
Comment (Lctrs) [reply com.microsoft.teams.Teams.json]: You can also upload files from your computer. Limiting it to downloads is too restrictive for a good user experience.

Comment (barthalion) [reply com.microsoft.teams.Teams.json]: It's easy to override permissions. We default to security where it's achievable.

File: com.microsoft.teams.Teams.json (hunk 6 above)
Comment (barthalion) [reply com.microsoft.teams.Teams.json]: It's not always about personal preferences but uniformity with existing applications.

**Comment by LeandroStanger** [comment]
bot, build com.microsoft.Teams

**Comment by Lctrs** [comment]
Grumpf. This appdata validation is too strict IMHO. I took the description from the official `.deb` and the caption of screenshots from the official [screenshots gallery](https://news.microsoft.com/microsoft-teams/#screenshots). I'll remove dots from the captions but not sure what I can do for the description...

File: com.microsoft.teams.Teams.json (hunk 6 above)
Comment (Lctrs) [reply com.microsoft.teams.Teams.json]: For the record, I strongly disagree with you.

But that's not worth the hassle. I'll just do it for the `ar` and `desktop-file-edit` modules, but not the shell scripts. Lot of existing applications also have them separated.

**Comment by LeandroStanger** [comment]
bot, build com.microsoft.Teams

File: apply_extra.sh [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+set -e
+
+ar x teams.deb
+tar xf data.tar.xz
```
Comment (gasinvein) [inline apply_extra.sh]: This can be done at once via pipe, like this: `ar p teams.deb data.tar.xz | tar -xJf -`, thus reducing IO and used space usage.

Comment (Lctrs) [reply apply_extra.sh]: Neat.

File: teams.sh (hunk 4 above)
Comment (gasinvein) [reply teams.sh]: Setting `TMDPR` via finish-args would allow you to remove the wrapper altogether.

**Comment by LeandroStanger** [comment]
bot, build com.microsoft.Teams

File: teams.sh (hunk 4 above)
Comment (Lctrs) [reply teams.sh]: The wrapper script is required for several reasons : 

1. Flatpak will error the build if it can't find the command defined, which will be the case since the binary called in the wrapper is in the extra data.
2. The binary in the extra data is just a shell script that expects to launch `../share/teams/teams`. Due to flatpak using read only filesystems, I can't put their binary inside `/app/bin` and the rest inside `/app/share`.

Also the wrapper script have the advantage to use an env var and to not let the users override this since it is required.

Comment (gasinvein) [reply teams.sh]: >Flatpak will error the build if it can't find the command defined

You can add a symlink (`/app/bin/teams` → `../extra/share/teams/teams` or something).

Comment (Lctrs) [reply teams.sh]: Done.

**Comment by LeandroStanger** [comment]
bot, build com.microsoft.Teams

**Comment by nedrichards** [comment]
@Lctrs Maybe look at the description for the android app? That's a bit more compact and 'app store ready'.

**Comment by Lctrs** [comment]
@nedrichards Great idea ! Here you go.

**Comment by Lctrs** [comment]
bot, build com.microsoft.teams.Teams

**Comment by Lctrs** [comment]
bot, build com.microsoft.Teams

**Comment by LeandroStanger** [comment]
bot, build com.microsoft.Teams

File: apply_extra.sh (hunk 9 above)
Comment (Lctrs) [reply apply_extra.sh]: @barthalion Do you still require that I inline this file ?

File: com.microsoft.teams.Teams.json (hunk 8 above)
Comment (Lctrs) [reply com.microsoft.teams.Teams.json]: I'll try to come up with something as well if I found time.

File: apply_extra.sh (hunk 9 above)
Comment (barthalion) [reply apply_extra.sh]: It's not a hill I'm willing to die on.

File: com.microsoft.teams.Teams.json (hunk 5 above)
Comment (Lctrs) [reply com.microsoft.teams.Teams.json]: @barthalion I'm back at it again. [Is doing the same as the slack wrapper acceptable ?](https://github.com/flathub/com.slack.Slack/blob/0650ab1f3cc6c9f5425fa74f8de565fe39aab7b1/com.slack.Slack.json#L16)

Comment (barthalion) [reply com.microsoft.teams.Teams.json]: Yes, that would be better.

File: com.microsoft.teams.Teams.json (hunk 8 above)
Comment (Lctrs) [reply com.microsoft.teams.Teams.json]: Please see https://github.com/flathub/flatpak-external-data-checker/pull/61
